wgpu/src/context.rs

```diff
@@ -0,0 +1,3873 @@
+        pass_data: &mut Self::RenderPassData,
+        group_label: &str,
+    );
+    fn render_pass_pop_debug_group(
+        &self,
+        pass: &mut Self::RenderPassId,
+        pass_data: &mut Self::RenderPassData,
+    );
+    fn render_pass_write_timestamp(
+        &self,
+        pass: &mut Self::RenderPassId,
+        pass_data: &mut Self::RenderPassData,
+        query_set: &Self::QuerySetId,
+        query_set_data: &Self::QuerySetData,
+        query_index: u32,
+    );
+    fn render_pass_begin_pipeline_statistics_query(
+        &self,
+        pass: &mut Self::RenderPassId,
+        pass_data: &mut Self::RenderPassData,
+        query_set: &Self::QuerySetId,
+        query_set_data: &Self::QuerySetData,
+        query_index: u32,
+    );
+    fn render_pass_end_pipeline_statistics_query(
+        &self,
+        pass: &mut Self::RenderPassId,
+        pass_data: &mut Self::RenderPassData,
+    );
+    fn render_pass_execute_bundles<'a>(
+        &self,
+        pass: &mut Self::RenderPassId,
+        pass_data: &mut Self::RenderPassData,
+        render_bundles: Box<dyn Iterator<Item = Self::RenderBundleId> + 'a>,
+    );
+}
+
+/// Object id.
+///
+/// An ObjectId is a 128-bit number internally, where the first 64-bits represent a backend internal id and
+/// the last 64-bits are a global id.
+#[derive(Debug, Clone, Copy)]
+pub struct ObjectId(NonZeroU128);
+
+impl ObjectId {
+    // DO NOT MERGE: This is definitely not ideal
+    pub fn dummy() -> Self {
+        Self(NonZeroU128::new(u128::MAX).unwrap())
+    }
+
+    pub fn global_id(&self) -> u64 {
+        (self.0.get() >> 64) as u64
+    }
+}
+
+impl From<NonZeroU128> for ObjectId {
+    fn from(raw: NonZeroU128) -> Self {
+        Self(raw)
+    }
+}
+
+impl From<ObjectId> for NonZeroU128 {
+    fn from(id: ObjectId) -> Self {
+        id.0
+    }
+}
+
+static_assertions::assert_impl_all!(ObjectId: Send, Sync);
+
+fn downcast_ref<T: Debug + Send + Sync + 'static>(data: &crate::Data) -> &T {
+    strict_assert!(data.is::<T>());
+    // Copied from std.
+    unsafe { &*(data as *const dyn Any as *const T) }
+}
+
+fn downcast_mut<T: Debug + Send + Sync + 'static>(data: &mut crate::Data) -> &mut T {
+    strict_assert!(data.is::<T>());
+    // Copied from std.
+    unsafe { &mut *(data as *mut dyn Any as *mut T) }
+}
+
+/// An object safe variant of [`Context`] implemented by all types that implement [`Context`].
+pub(crate) trait DynContext: Debug + Send + Sync {
+    fn as_any(&self) -> &dyn Any;
+
+    fn instance_create_surface(
+        &self,
+        display_handle: raw_window_handle::RawDisplayHandle,
+        window_handle: raw_window_handle::RawWindowHandle,
+    ) -> (ObjectId, Box<crate::Data>);
+    #[allow(clippy::type_complexity)]
+    fn instance_request_adapter(
+        &self,
+        options: &RequestAdapterOptions<'_>,
+    ) -> Pin<Box<dyn Future<Output = Option<(ObjectId, Box<crate::Data>)>> + Send>>;
+    #[allow(clippy::type_complexity)]
```

Comment:
Okay I think this one is justified lol

Could we figure out how to simplify this.